### PR TITLE
doc: change the branching strategy

### PR DESCRIPTION
**Pull request by patricklodder**
Major administrative policy change: change the branching strategy to be able to release software for multiple versions asynchronously, while having an integral up to date but unstable master branch.

Soon, a 1.14.8 minor release will have to be done, but the major version being worked on right now - 1.15.0 - is not ready to be released around the same time. This means that under the current branching strategy, there will either be a `master` branch that does not contain fixes from 1.14.8, or one that cannot be cleanly fast-forwarded into 1.15.0. Both outcomes are undesirable.

The fix is easy: since patches apply to both `1.15.0` and `1.14.8` (almost 1-on-1), by making `master` unstable and tracking the latest work there. A major release (other than a re-port from upstream) becomes simply a copy of master when creating a development branch at the time of releasing a candidate, and a minor release is - as is now - a copy of the maintenance branch, which can then get the desired commits picked into them. This is largely the upstream branching strategy. If another full re-port is required (unforeseen at this time), this can happen on a separate `-dev` branch (and will disrupt `master` the same way it did in the past.

Practically this means:

- All pull requests that are not release-specific will need to be done to `master` - every new feature or bug fix needs to go there first.
- When self-compiling, `master` is no longer a safe choice for production deployments - tags need to be checked out.
- Maintainers will have a little more work when preparing releases, as sometimes, bug fixes need to be backported to multiple branches.

Additional upsides:
- Documentation and tooling on `master` will always be the latest version

-----

Administrative note: please discuss, but do not merge this yet. We need to have a decision towards this on the day that 1.14.8 gets released, before that, it'd be premature to change the documented policy.

### CODE REVIEW

**Comment by victorsk2019** [comment]
Hi.  To clarify, you are proposing making `master` branch more of a "rolling" branch, yes?

**Comment by patricklodder** [comment]
> To clarify, you are proposing making master branch more of a "rolling" branch, yes?

Yes. About 10 years ago, we changed it from that, to accommodate full back ports of Bitcoin Core better. Now, that makes much less sense, so I'm basically proposing to change it back to master being the integration branch, simply because right now, 1.15.0-dev is technically functioning as the integration brach, but as described above, I foresee issues when we're releasing a 1.14.8 but not a 1.15.0 version.

**Comment by chromatic** [comment]
My only concern is that this adds some maintenance burden to cherry-pick into a hypothetical 1.14.8 and 1.15.0 simultaneously.

However, we maintainers can reduce this burden by not attempting to maintain two branches simultaneously , which I believe is the intent: branch off a 1.14.8 release and prepare and fixate it there, then merge back anything essential into the main branch.

Thus we'll only pay that multi-branch cost if we're preparing for two simultaneous releases, and that's almost entirely under our control.

**Comment by patricklodder** [comment]
> My only concern is that this adds some maintenance burden to cherry-pick into a hypothetical 1.14.8 and 1.15.0 simultaneously.

If we now merge 1.15.0-dev into master, it means that when we push a new 1.15.0-dev close to release, i.e. for an RC, then we'd only have to cherry-pick those items that come in after, and maybe revert any commits that should be omitted. So the impact should be minimal.

For 1.14.8 we anyway have to do this, no matter what. Same as we did for 1.8.3 (and were supposed to do for 1.10.1), once upon a time.

**Comment by patricklodder** [comment]
>  branch off a 1.14.8 release and prepare and fixate it there, then merge back anything essential into the main branch.

Minor correction: we branch off a 1.14.8 and then maintainers cherry-pick into that, only after something got merged to the main branch. So we make it one-way. We can have multiple of those at the time, but if they all follow this pattern, then we never get into conflicts between what is and what is not staged in the main branch, which is our current situation.

**Review by Skandiakiwi13 — APPROVED** [review]

**Review by Skandiakiwi13 — APPROVED** [review]
Awesome work

**Comment by chromatic** [comment]
We've branched off for the 1.14.8 release and cherry-picked from 1.15.0-dev into it, so I think we've shown that this process *can* work even if we're not following it to the letter. With this and the suggestion of a `README.md` update (@daank-c you should open a PR for that!), I'm :+1: on the change.

**Comment by patricklodder** [comment]
I've added a warning to the top of the document.

**Comment by patricklodder** [comment]
This is now urgent, as 1.14.8 is out and master does not include those bugfixes.

**Comment by chromatic** [comment]
Good to go with a squash; I'll approve once that's in place and we can merge one hour after approval.

**Comment by patricklodder** [comment]
squash done

**Pull request merged**
